src/Ztm.WebApi/AddressPools/ReceivingAddressPool.cs

```diff
@@ -37,81 +34,27 @@ public sealed class ReceivingAddressPool : IReceivingAddressPool
         {
-            storageLock.EnterWriteLock();
-
-            try
-            {
                 var address = await this.generator.GenerateAsync(cancellationToken);
-                await this.storage.AddAddressAsync(address, cancellationToken);
-            }
-            finally
-            {
-                storageLock.ExitWriteLock();
-            }
+                await this.storage.AddAddressAsync(address, CancellationToken.None);
         }
 
-        public async Task ReleaseAsync(Guid id, CancellationToken cancellationToken)
+        public Task ReleaseAddressAsync(Guid id, CancellationToken cancellationToken)
         {
-            this.storageLock.EnterWriteLock();
-
-            try
-            {
-                var r = await this.storage.GetReservationAsync(id, cancellationToken);
-                if (r == null)
-                {
-                    throw new KeyNotFoundException("The reservation is not found.");
-                }
-
-                if (r.ReleasedDate != null)
-                {
-                    throw new InvalidOperationException("The reservation is already released.");
-                }
-
-                await this.storage.SetLockedStatusAsync(r.ReceivingAddress.Id, false, cancellationToken);
-                await this.storage.SetReleasedTimeAsync(id, cancellationToken);
-            }
-            finally
-            {
-                this.storageLock.ExitWriteLock();
-            }
+            return this.ReleaseAddressAsync(id, cancellationToken);
         }
 
         public async Task<ReceivingAddressReservation> TryLockAddressAsync(CancellationToken cancellationToken)
         {
-            this.storageLock.EnterWriteLock();
-
-            try
-            {
-                var addresses = await this.storage.ListReceivingAddressAsync(cancellationToken);
-                if (addresses != null)
-                {
-                    var availables = addresses.Where(a => a.Available);
-
-                    if (availables.Count() <= 0)
-                    {
-                        return null;
-                    }
-
-                    var chosen = this.choser.Choose(availables);
+            var addresses = await this.storage.ListReceivingAddressAsync(AddressFilter.Available, cancellationToken);
 
-                    if (chosen != null)
-                    {
-                        var reservation = await this.storage.CreateReservationAsync(chosen.Id, cancellationToken);
-                        await this.storage.SetLockedStatusAsync(chosen.Id, true, cancellationToken);
-
-                        return reservation;
-                    }
-                }
-            }
-            finally
+            if (addresses.Any())
             {
-                this.storageLock.ExitWriteLock();
+                var chosen = this.choser.Choose(addresses);
+                return await this.storage.TryLockAsync(chosen.Id, CancellationToken.None);
```

Comment:
We can allow to cancel here due to all of above operations does not cause any state to change.